core/server/master/src/main/java/alluxio/master/throttle/SystemMonitor.java

```diff
@@ -0,0 +1,391 @@
+        mPrevPitInfo != null ? mPrevPitInfo.getJVMPauseTime() : 0);
+    serverIndicator.setPitTimeMS(snapshotTimeMS);
+
+    mServerIndicatorsList.add(serverIndicator);
+    if (mServerIndicatorsList.size() > mMaxNumberOfSnapshot && !mServerIndicatorsList.isEmpty()) {
+      ServerIndicator toBeRemoved = mServerIndicatorsList.get(0);
+      if (mAggregatedServerIndicators != null) {
+        mAggregatedServerIndicators.reduction(toBeRemoved);
+      }
+      mServerIndicatorsList.remove(0);
+    }
+
+    if (mAggregatedServerIndicators == null) {
+      mAggregatedServerIndicators = new ServerIndicator(serverIndicator);
+    } else {
+      mAggregatedServerIndicators.addition(serverIndicator);
+    }
+  }
+
+  private StatusTransition statusCheck(ServerIndicator aggregatedServerIndicator,
+      ServerIndicator pitServerIndicator,
+      ServerIndicator lowBoundaryPitIndicator,
+      ServerIndicator highBoundaryPitIndicator,
+      ServerIndicator lowBoundaryAggregateIndicator,
+      ServerIndicator highBoundaryAggregateIndicator) {
+    // Just start with simple.
+    // When it is clear, more rules can be added
+
+    // The pitServerIndicator shows the current status, it reflects the current status.
+    // so it is checked first. If it is in reasonable range, the system should be
+    if (lowBoundaryPitIndicator != null
+        && pitServerIndicator.getHeapUsed() < lowBoundaryPitIndicator.getHeapUsed()) {
+      return StatusTransition.DEESCALATE;
+    }
+
+    if (highBoundaryPitIndicator != null
+        && pitServerIndicator.getHeapUsed() < highBoundaryPitIndicator.getHeapUsed()) {
+      // if the heap usage is not crossing the higher boundary, no change
+      return StatusTransition.UNCHANGED;
+    }
+
+    if (highBoundaryAggregateIndicator != null
+        && aggregatedServerIndicator.getHeapUsed()
+        > highBoundaryAggregateIndicator.getHeapUsed()) {
+      return StatusTransition.ESCALATE;
+    }
+
+    return StatusTransition.UNCHANGED;
+  }
+
+  private void checkAndBackPressure() {
+    if (mServerIndicatorsList.isEmpty()) {
+      return;
+    }
+
+    reInitTheThresholds();
+
+    ServerIndicator pitIndicator = mServerIndicatorsList.get(mServerIndicatorsList.size() - 1);
+    StatusTransition statusTransition = StatusTransition.UNCHANGED;
+
+    switch (mCurrentSystemStatus) {
+      case IDLE:
+        statusTransition = statusCheck(mAggregatedServerIndicators, pitIndicator, null,
+            mPitThresholdActive, null, mAggregateThresholdActive);
+        if (statusTransition == StatusTransition.ESCALATE) {
+          mCurrentSystemStatus = SystemStatus.ACTIVE;
+        }
+        break;
+      case ACTIVE:
+        statusTransition = statusCheck(mAggregatedServerIndicators, pitIndicator,
+            mPitThresholdActive, mPitThresholdStressed,
+            mAggregateThresholdActive, mAggregateThresholdStressed);
+        if (statusTransition == StatusTransition.DEESCALATE) {
```

Comment:
Maybe add a function that can do the escalate/descalate as part of the SystemStats class to make this code cleaner?

Reply:
I will check that in next pr. still considering whether to add the pause time as the factor.
Eg, previous heartbeat happens at T, while the next heartbeat happens at T + 10 (suppose to be T + 3), Or T + 60